File: move_zeroes.c
/*

283. Move Zeroes

Question: https://leetcode.com/problems/move-zeroes/

*/

/*
	Easy to understand solution
*/

void moveZeroes(int* nums, int numsSize) {
    int i,j;
    
    for(i = 0; i < numsSize; i++){
        
        if( nums[i] == 0 ){
            
            for(j = i+1; j < numsSize; j++){
                
                if(nums[j] != 0){
                    nums[i] = nums[j];
                    nums[j] = 0;
                    break;
                }

            }

        }

    }
}



/*
	Optimal Solution
*/

void moveZeroes(int* nums, int numsSize) {
    int i, numZeroes = 0;
    for(i = 0; i < numsSize; i++){
        if( nums[i] == 0 ){
            numZeroes++;
        }else{
            nums[i - numZeroes] = nums[i];
        }
    }
    while(numZeroes > 0){
        nums[numsSize - numZeroes] = 0;
        numZeroes--;
    }
}

File: move_zeroes.py
"""
283. Move Zeroes

Question: https://leetcode.com/problems/move-zeroes/

"""

"""
Easy to understand solution
"""

class Solution(object):
    def moveZeroes(self, nums):
        """
        :type nums: List[int]
        :rtype: void Do not return anything, modify nums in-place instead.
        """
        lenn = len(nums)
        for i in range(lenn):
            if nums[i] == 0:
                for itr in range(i+1,lenn):
                    if nums[itr] != 0:
                        t = nums[itr]
                        nums[itr] = nums[i]
                        nums[i] = t
                        break

"""
Optimal solution
"""

class Solution(object):
    def moveZeroes(self, nums):
        """
        :type nums: List[int]
        :rtype: void Do not return anything, modify nums in-place instead.
        """
        numZeroes = 0
        lenn = len(nums)
        for i in range(lenn):
            if nums[i] == 0:
                numZeroes += 1
            else:
                nums[i - numZeroes] = nums[i]
        while numZeroes > 0:
            nums[lenn-numZeroes] = 0
            numZeroes -= 1